# Employee Performance Prediction

This notebook uses the trained XGBoost pipeline to predict the PerformanceRating of an employee based on the input employee attributes.

The trained model was developed in `train_model.ipynb` and saved as a complete preprocessing and prediction pipeline.

The prediction workflow consists of:

1. Loading the saved model
2. Preparing employee input data
3. Generating the predicted PerformanceRating
4. Displaying prediction probabilities
5. Interpreting the prediction result

The model uses the same feature structure and preprocessing steps applied during model training.

## 3.1 Import Required Libraries

The required libraries are imported for loading the saved model, creating employee input data, and generating predictions.

In [1]:
import os
import joblib
import pandas as pd

## 3.2 Load the Trained Model

The final tuned XGBoost pipeline saved during the model training stage is loaded from the project `models` directory.

The complete pipeline is loaded so that the same preprocessing used during training is automatically applied to new employee data.

In [2]:
# ============================================================
# Load Final Trained Model
# ============================================================

model_path = "../../models/final_xgboost_employee_performance_model.joblib"

model = joblib.load(model_path)

print("=" * 80)
print("FINAL MODEL LOADED")
print("=" * 80)

print(f"\nModel Type : {type(model).__name__}")
print(f"Model Path : {model_path}")
print(f"File Exists: {os.path.exists(model_path)}")

FINAL MODEL LOADED

Model Type : Pipeline
Model Path : ../../models/final_xgboost_employee_performance_model.joblib
File Exists: True


## 3.3 Define Prediction Features

The trained model was developed using 26 employee attributes as predictors.

The prediction input must contain the same feature names and structure used during model training. The employee identifier `EmpNumber` and target variable `PerformanceRating` are not included because the identifier was excluded from modeling and the target is the value being predicted.

In [4]:
# ============================================================
# Define Prediction Features
# ============================================================

prediction_features = [
    "Age",
    "Gender",
    "EducationBackground",
    "MaritalStatus",
    "EmpDepartment",
    "EmpJobRole",
    "BusinessTravelFrequency",
    "DistanceFromHome",
    "EmpEducationLevel",
    "EmpEnvironmentSatisfaction",
    "EmpHourlyRate",
    "EmpJobInvolvement",
    "EmpJobLevel",
    "EmpJobSatisfaction",
    "NumCompaniesWorked",
    "OverTime",
    "EmpLastSalaryHikePercent",
    "EmpRelationshipSatisfaction",
    "TotalWorkExperienceInYears",
    "TrainingTimesLastYear",
    "EmpWorkLifeBalance",
    "ExperienceYearsAtThisCompany",
    "ExperienceYearsInCurrentRole",
    "YearsSinceLastPromotion",
    "YearsWithCurrManager",
    "Attrition"
]

print("=" * 80)
print("PREDICTION FEATURES")
print("=" * 80)

print(f"\nNumber of prediction features: {len(prediction_features)}")

print("\nFeatures:")
for i, feature in enumerate(prediction_features, start=1):
    print(f"{i:2}. {feature}")

PREDICTION FEATURES

Number of prediction features: 26

Features:
 1. Age
 2. Gender
 3. EducationBackground
 4. MaritalStatus
 5. EmpDepartment
 6. EmpJobRole
 7. BusinessTravelFrequency
 8. DistanceFromHome
 9. EmpEducationLevel
10. EmpEnvironmentSatisfaction
11. EmpHourlyRate
12. EmpJobInvolvement
13. EmpJobLevel
14. EmpJobSatisfaction
15. NumCompaniesWorked
16. OverTime
17. EmpLastSalaryHikePercent
18. EmpRelationshipSatisfaction
19. TotalWorkExperienceInYears
20. TrainingTimesLastYear
21. EmpWorkLifeBalance
22. ExperienceYearsAtThisCompany
23. ExperienceYearsInCurrentRole
24. YearsSinceLastPromotion
25. YearsWithCurrManager
26. Attrition


## 3.4 Create Sample Employee Input

A sample employee record is created using the same 26 predictor variables used during model training.

The sample values represent a hypothetical employee and are provided only to demonstrate how the trained model can be used for prediction.

The target variable `PerformanceRating` is not provided because it is the value to be predicted.

In [5]:
# ============================================================
# Create Sample Employee Input
# ============================================================

sample_employee = pd.DataFrame([{
    "Age": 35,
    "Gender": "Male",
    "EducationBackground": "Life Sciences",
    "MaritalStatus": "Married",
    "EmpDepartment": "Development",
    "EmpJobRole": "Developer",
    "BusinessTravelFrequency": "Travel_Rarely",
    "DistanceFromHome": 5,
    "EmpEducationLevel": 3,
    "EmpEnvironmentSatisfaction": 4,
    "EmpHourlyRate": 65,
    "EmpJobInvolvement": 3,
    "EmpJobLevel": 2,
    "EmpJobSatisfaction": 3,
    "NumCompaniesWorked": 2,
    "OverTime": "No",
    "EmpLastSalaryHikePercent": 18,
    "EmpRelationshipSatisfaction": 3,
    "TotalWorkExperienceInYears": 10,
    "TrainingTimesLastYear": 3,
    "EmpWorkLifeBalance": 3,
    "ExperienceYearsAtThisCompany": 6,
    "ExperienceYearsInCurrentRole": 4,
    "YearsSinceLastPromotion": 1,
    "YearsWithCurrManager": 3,
    "Attrition": "No"
}])

print("=" * 80)
print("SAMPLE EMPLOYEE INPUT")
print("=" * 80)

print(f"\nInput shape: {sample_employee.shape}")

display(sample_employee)

SAMPLE EMPLOYEE INPUT

Input shape: (1, 26)


,Age,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,...,EmpLastSalaryHikePercent,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,Attrition
0,35,Male,Life Sciences,Married,Development,Developer,Travel_Rarely,5,3,4,...,18,3,10,3,3,6,4,1,3,No


## 3.5 Predict Employee Performance Rating

The trained XGBoost pipeline is used to predict the PerformanceRating for the sample employee.

Because the target variable contains the original project ratings 2, 3, and 4, the prediction will also be expressed using these rating values.

In [7]:
# ============================================================
# Generate Performance Rating Prediction
# ============================================================

# XGBoost uses internally encoded labels:
# 0 -> PerformanceRating 2
# 1 -> PerformanceRating 3
# 2 -> PerformanceRating 4

xgb_label_mapping = {
    0: 2,
    1: 3,
    2: 4
}

predicted_internal_label = model.predict(sample_employee)[0]

predicted_rating = xgb_label_mapping[int(predicted_internal_label)]

print("=" * 80)
print("EMPLOYEE PERFORMANCE PREDICTION")
print("=" * 80)

print(f"\nInternal Model Prediction : {predicted_internal_label}")
print(f"Predicted Performance Rating: {predicted_rating}")

EMPLOYEE PERFORMANCE PREDICTION

Internal Model Prediction : 1
Predicted Performance Rating: 3


## 3.6 Prediction Probabilities

The trained model provides probabilities for each of its internal classes.

These probabilities are mapped back to the original PerformanceRating values used in the INX Future Inc. dataset (2, 3, and 4).

The class with the highest probability is selected as the predicted PerformanceRating.

In [8]:
# ============================================================
# Prediction Probabilities
# ============================================================

prediction_probabilities = model.predict_proba(sample_employee)[0]

probability_mapping = pd.DataFrame({
    "PerformanceRating": [2, 3, 4],
    "Probability": prediction_probabilities
})

probability_mapping["Probability (%)"] = (
    probability_mapping["Probability"] * 100
).round(2)

print("=" * 80)
print("PREDICTION PROBABILITIES")
print("=" * 80)

display(probability_mapping)

PREDICTION PROBABILITIES


,PerformanceRating,Probability,Probability (%)
0,2,0.002353,0.240000
1,3,0.985518,98.550003
2,4,0.012130,1.210000


## 3.7 Reusable Employee Performance Prediction Function

A reusable prediction function is created to generate the PerformanceRating and corresponding class probabilities for a new employee.

The function accepts an employee record containing the 26 predictor variables and applies the saved preprocessing and XGBoost model automatically.

In [9]:
# ============================================================
# Reusable Employee Performance Prediction Function
# ============================================================

def predict_employee_performance(employee_data, model):
    """
    Predict PerformanceRating for a new employee.

    Parameters
    ----------
    employee_data : pandas.DataFrame
        Employee data containing the 26 predictor variables.
    model : sklearn Pipeline
        Trained employee performance prediction pipeline.

    Returns
    -------
    predicted_rating : int
        Predicted PerformanceRating (2, 3, or 4).
    probability_table : pandas.DataFrame
        Probability for each PerformanceRating.
    """

    # Generate internal XGBoost prediction
    internal_prediction = int(model.predict(employee_data)[0])

    # Convert internal XGBoost class to original project rating
    xgb_label_mapping = {
        0: 2,
        1: 3,
        2: 4
    }

    predicted_rating = xgb_label_mapping[internal_prediction]

    # Generate probabilities
    probabilities = model.predict_proba(employee_data)[0]

    probability_table = pd.DataFrame({
        "PerformanceRating": [2, 3, 4],
        "Probability (%)": (probabilities * 100).round(2)
    })

    return predicted_rating, probability_table

## 3.8 Test the Prediction Function

The reusable prediction function is tested using the sample employee created earlier.

The prediction should be consistent with the result obtained directly from the trained pipeline.

In [10]:
# ============================================================
# Test Prediction Function
# ============================================================

predicted_rating, probability_table = predict_employee_performance(
    sample_employee,
    model
)

print("=" * 80)
print("REUSABLE PREDICTION FUNCTION RESULT")
print("=" * 80)

print(f"\nPredicted Performance Rating: {predicted_rating}")

print("\nPrediction Probabilities:")
display(probability_table)

REUSABLE PREDICTION FUNCTION RESULT

Predicted Performance Rating: 3

Prediction Probabilities:


,PerformanceRating,Probability (%)
0,2,0.240000
1,3,98.550003
2,4,1.210000


## 3.9 Validate Prediction Input

Before generating predictions for new employees, the input structure is checked against the expected prediction features.

This ensures that all required predictor variables are present and that no unexpected variables are passed to the model.

In [11]:
# ============================================================
# Validate Prediction Input
# ============================================================

missing_features = set(prediction_features) - set(sample_employee.columns)
extra_features = set(sample_employee.columns) - set(prediction_features)

print("=" * 80)
print("PREDICTION INPUT VALIDATION")
print("=" * 80)

print(f"\nExpected features : {len(prediction_features)}")
print(f"Input features    : {len(sample_employee.columns)}")

print(f"\nMissing features  : {missing_features}")
print(f"Extra features    : {extra_features}")

if not missing_features and not extra_features:
    print("\nValidation Status : PASSED")
else:
    print("\nValidation Status : FAILED")

PREDICTION INPUT VALIDATION

Expected features : 26
Input features    : 26

Missing features  : set()
Extra features    : set()

Validation Status : PASSED


## 3.10 Prediction Summary

For the sample employee, the trained XGBoost model predicts a `PerformanceRating` of **3**.

The model assigns the highest probability to Rating 3, indicating that Rating 3 is the most likely predicted performance category for this sample employee.

This prediction demonstrates how the saved model can be used to estimate employee performance from the available input attributes.

The prediction is model-based and should be interpreted as an estimate rather than a guarantee of future employee performance.

In [12]:
print("=" * 80)
print("PREDICTION SUMMARY")
print("=" * 80)

print(f"\nPredicted Performance Rating: {predicted_rating}")

highest_probability = probability_table.loc[
    probability_table["Probability (%)"].idxmax()
]

print(
    f"Highest Probability Rating : "
    f"{highest_probability['PerformanceRating']}"
)

print(
    f"Highest Probability        : "
    f"{highest_probability['Probability (%)']:.2f}%"
)

PREDICTION SUMMARY

Predicted Performance Rating: 3
Highest Probability Rating : 3.0
Highest Probability        : 98.55%


## 3.11 Predict Performance for Multiple Employees

The trained model can generate predictions for multiple employee records at the same time.

This demonstrates how the prediction pipeline can be applied to a batch of employee inputs rather than only a single employee record.

The same 26 predictor variables are used for each employee, and the predicted PerformanceRating is returned for every record.

In [13]:
# ============================================================
# Multiple Employee Prediction
# ============================================================

multiple_employees = pd.concat(
    [
        sample_employee,
        sample_employee.copy()
    ],
    ignore_index=True
)

# Modify a few values for the second hypothetical employee
multiple_employees.loc[1, "EmpEnvironmentSatisfaction"] = 2
multiple_employees.loc[1, "EmpLastSalaryHikePercent"] = 12
multiple_employees.loc[1, "OverTime"] = "Yes"

# Generate predictions
multiple_predictions = model.predict(multiple_employees)

# Convert internal XGBoost labels to original PerformanceRating
multiple_predictions = [
    xgb_label_mapping[int(prediction)]
    for prediction in multiple_predictions
]

prediction_results = multiple_employees.copy()
prediction_results["PredictedPerformanceRating"] = multiple_predictions

print("=" * 80)
print("MULTIPLE EMPLOYEE PREDICTIONS")
print("=" * 80)

display(
    prediction_results[
        [
            "EmpDepartment",
            "EmpJobRole",
            "EmpEnvironmentSatisfaction",
            "EmpLastSalaryHikePercent",
            "OverTime",
            "PredictedPerformanceRating"
        ]
    ]
)

MULTIPLE EMPLOYEE PREDICTIONS


,EmpDepartment,EmpJobRole,EmpEnvironmentSatisfaction,EmpLastSalaryHikePercent,OverTime,PredictedPerformanceRating
0,Development,Developer,4,18,No,3
1,Development,Developer,2,12,Yes,3


## 3.12 Batch Prediction with Class Probabilities

In addition to the predicted PerformanceRating, the model can provide the probability associated with each possible rating.

This allows the prediction results to show not only the predicted class but also the model's confidence distribution across Ratings 2, 3, and 4.

In [14]:
# ============================================================
# Batch Prediction with Probabilities
# ============================================================

batch_probabilities = model.predict_proba(multiple_employees)

batch_probability_results = pd.DataFrame({
    "PredictedPerformanceRating": multiple_predictions,
    "Rating_2_Probability_%": (batch_probabilities[:, 0] * 100).round(2),
    "Rating_3_Probability_%": (batch_probabilities[:, 1] * 100).round(2),
    "Rating_4_Probability_%": (batch_probabilities[:, 2] * 100).round(2)
})

print("=" * 80)
print("BATCH PREDICTION WITH PROBABILITIES")
print("=" * 80)

display(batch_probability_results)

BATCH PREDICTION WITH PROBABILITIES


,PredictedPerformanceRating,Rating_2_Probability_%,Rating_3_Probability_%,Rating_4_Probability_%
0,3,0.24,98.550003,1.21
1,3,7.11,92.050003,0.84


## 3.13 Prediction Notebook Conclusion

The trained XGBoost model was successfully loaded and used to predict employee PerformanceRating.

The prediction workflow demonstrated:

- Loading the saved trained model pipeline
- Preparing employee inputs using the 26 predictor variables
- Predicting PerformanceRating values of 2, 3, or 4
- Obtaining probability estimates for each performance category
- Creating a reusable prediction function
- Validating the prediction input structure
- Generating predictions for multiple employee records

For the demonstrated sample employee, the model predicted PerformanceRating 3 with a probability of 98.55%.

The model provides a data-driven estimate of employee performance based on the patterns learned from the available employee dataset. These predictions should support decision-making rather than be treated as a guarantee of future employee performance.